prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,178 @@
+type GCSArtifactFetcher struct {
+	client      *storage.Client
+	xmlEndpoint string
+	withTLS     bool
+}
+
+// gcsJobSource is a location in GCS where Prow job-specific artifacts are stored. This implementation assumes
+// Prow's native GCS upload format (treating GCS keys as a directory structure), and is not
+// intended to support arbitrary GCS bucket upload formats.
+type gcsJobSource struct {
+	source     string
+	linkPrefix string
+	bucket     string
+	jobPrefix  string
+	jobName    string
+	buildID    string
+}
+
+// NewGCSArtifactFetcher creates a new ArtifactFetcher with a real GCS Client
+func NewGCSArtifactFetcher(c *storage.Client, xmlEndpoint string, tls bool) *GCSArtifactFetcher {
+	return &GCSArtifactFetcher{
+		client:      c,
+		xmlEndpoint: xmlEndpoint,
+		withTLS:     tls,
+	}
+}
+
+func fieldsForJob(src jobSource) logrus.Fields {
+	return logrus.Fields{
+		"jobPrefix": src.jobPath(),
+	}
+}
+
+// newGCSJobSource creates a new gcsJobSource from a given bucket and jobPrefix
+func newGCSJobSource(src string) (*gcsJobSource, error) {
+	gcsURL, err := url.Parse(src)
+	if err != nil {
+		return &gcsJobSource{}, ErrCannotParseSource
+	}
+	gcsPath := &gcs.Path{}
+	err = gcsPath.SetURL(gcsURL)
+	if err != nil {
+		return &gcsJobSource{}, ErrCannotParseSource
+	}
+
+	tokens := strings.FieldsFunc(gcsPath.Object(), func(c rune) bool { return c == '/' })
+	if len(tokens) < 2 {
+		return &gcsJobSource{}, ErrCannotParseSource
+	}
+	buildID := tokens[len(tokens)-1]
+	name := tokens[len(tokens)-2]
+	return &gcsJobSource{
+		source:     src,
+		linkPrefix: "gs://",
+		bucket:     gcsPath.Bucket(),
+		jobPrefix:  gcsPath.Object() + "/",
+		jobName:    name,
+		buildID:    buildID,
+	}, nil
+}
+
+// Artifacts lists all artifacts available for the given job source
+func (af *GCSArtifactFetcher) artifacts(src jobSource) []string {
+	listStart := time.Now()
+	bucketName, prefix := extractBucketPrefixPair(src.jobPath())
+	artifacts := []string{}
+	bkt := af.client.Bucket(bucketName)
+	q := storage.Query{
+		Prefix:   prefix,
+		Versions: false,
+	}
+	objIter := bkt.Objects(context.Background(), &q)
+	for {
+		oAttrs, err := objIter.Next()
+		if err != nil {
+			if err == iterator.Done {
+				break
+			}
+			logrus.WithFields(fieldsForJob(src)).Error("Error accessing GCS artifact.")
+		}
+		artifacts = append(artifacts, strings.TrimPrefix(oAttrs.Name, prefix))
+
+	}
+	listElapsed := time.Since(listStart)
+	logrus.WithField("duration", listElapsed).Infof("Listed %d artifacts.", len(artifacts))
+	return artifacts
+}
+
+type gcsArtifactHandle struct {
+	*storage.ObjectHandle
+}
+
+func (h *gcsArtifactHandle) NewReader(ctx context.Context) (artifactReader, error) {
+	return h.ObjectHandle.NewReader(ctx)
+}
+
+func (h *gcsArtifactHandle) NewRangeReader(ctx context.Context, offset, length int64) (artifactReader, error) {
+	return h.ObjectHandle.NewRangeReader(ctx, offset, length)
```

Comment:
Sounds like this one will return a different error than what we're unit testing? Why isn't that a problem?

Reply:
addressed above, it will return the same